lib/semantics/resolve-names.cc

```diff
@@ -3568,6 +3571,78 @@ bool DeclarationVisitor::Pre(const parser::StructureConstructor &x) {
   return false;
 }
 
+bool DeclarationVisitor::Pre(const parser::BasedPointerStmt &x) {
+  for (const parser::BasedPointer &bp : x.v) {
+    const parser::ObjectName &pointerName{std::get<0>(bp.t)};
+    const parser::ObjectName &pointeeName{std::get<1>(bp.t)};
+    auto *pointer{FindSymbol(pointerName)};
+    if (!pointer) {
+      pointer = &MakeSymbol(pointerName, ObjectEntityDetails{});
+    } else if (pointer->has<UseDetails>()) {
+      Say(pointerName,
+          "'%s' cannot be a Cray pointer as it is use-associated"_err_en_US);
+    } else if (!ConvertToObjectEntity(*pointer) || IsParameter(*pointer)) {
+      SayWithDecl(pointerName, *pointer, "'%s' is not a variable"_err_en_US);
+    } else if (pointer->Rank() > 0) {
+      SayWithDecl(pointerName, *pointer,
+          "Cray pointer '%s' must be a scalar"_err_en_US);
+    } else if (pointer->test(Symbol::Flag::CrayPointee)) {
+      Say(pointerName,
+          "'%s' cannot be a Cray pointer as it is already a Cray pointee"_err_en_US);
+    }
+    pointer->set(Symbol::Flag::CrayPointer);
+    const DeclTypeSpec &pointerType{MakeNumericType(TypeCategory::Integer,
+        context().defaultKinds().subscriptIntegerKind())};
+    const auto *type{pointer->GetType()};
+    if (!type) {
+      pointer->SetType(pointerType);
+    } else if (*type != pointerType) {
+      Say(pointerName.source, "Cray pointer '%s' must have type %s"_err_en_US,
+          pointerName.source, pointerType.AsFortran());
+    }
+    if (ResolveName(pointeeName)) {
+      Symbol &pointee{*pointeeName.symbol};
+      if (pointee.has<UseDetails>()) {
+        Say(pointeeName,
+            "'%s' cannot be a Cray pointee as it is use-associated"_err_en_US);
+      } else if (!ConvertToObjectEntity(pointee) || IsParameter(pointee)) {
+        Say(pointeeName, "'%s' is not a variable"_err_en_US);
```

Comment:
Must pointees always have an `ObjectEntityDetails` ? Gfortran seems to allow pointee procedures, from the linked manual :
> A Cray pointer may also point to a function or a subroutine.